### ISSUE TITLE
_prepare_report_table: Inconsistencies in model_vars (BatchRunner vs BatchRunnerMP) lead to wrong 'Run' values

### ISSUE THREAD

**Issue by adrien-perello** [projectmesa/mesa#1077]
## Describe the bug
incorrect `Run` values when using `BatchRunnerMP`.  This seems to be due to assigning to `Run` the values of the first `fixed_parameters`

![Screenshot from 2021-11-13 16-18-30](https://user-images.githubusercontent.com/76055749/141649201-dd0d3e77-cf5a-41c7-b34f-0d7b177afefa.png)

## Expected behavior
The `Run` column should have the value of the appropriate iteration number

## To Reproduce
using the [BoltzmannWealthModel](https://github.com/projectmesa/mesa/tree/main/examples/boltzmann_wealth_model) (slightly modified to accept an `initial_wealth` parameter)

```python
batch_run = BatchRunnerMP(
    BoltzmannWealthModel,
    variable_parameters = {"wealth_ini": [1, 5, 10, 50, 100], 'N': [20,50]},
    fixed_parameters = {"width": 10, "height": 10},
    iterations = 5,
    max_steps = 10,
    model_reporters={"datacollectors" : lambda model: model.datacollector},
)

batch_run.run_all()
batch_run.get_model_vars_dataframe()
```

## The _prepare_report_table method
When inspecting the problem, it seems to have to do with the `_prepare_report_table` method in `batchrunner.py`:

```python
def _prepare_report_table(self, vars_dict, extra_cols=None):
    """
    Creates a dataframe from collected records and sorts it using 'Run'
    column as a key.
    """
    extra_cols = ["Run"] + (extra_cols or [])
    index_cols = []
    if self.parameters_list:
        index_cols = list(self.parameters_list[0].keys())
    index_cols += extra_cols

    records = []
    for param_key, values in vars_dict.items():
        # !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
        print(index_cols, param_key)  # PRINT FOR CHECKING
        # !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
        record = dict(zip(index_cols, param_key))
        record.update(values)
        records.append(record)
```

When I print the following `index_cols` and `param_key`, I get

```python
['wealth_ini', 'N', 'Run'] (1, 20, 10, 10, 1)
['wealth_ini', 'N', 'Run'] (1, 20, 10, 10, 0)
['wealth_ini', 'N', 'Run'] (1, 20, 10, 10, 3)
['wealth_ini', 'N', 'Run'] (1, 20, 10, 10, 4)
['wealth_ini', 'N', 'Run'] (1, 50, 10, 10, 0)
['wealth_ini', 'N', 'Run'] (5, 20, 10, 10, 1)
['wealth_ini', 'N', 'Run'] (5, 20, 10, 10, 2)
['wealth_ini', 'N', 'Run'] (1, 50, 10, 10, 4)
['wealth_ini', 'N', 'Run'] (1, 50, 10, 10, 3)
['wealth_ini', 'N', 'Run'] (5, 20, 10, 10, 0)
['wealth_ini', 'N', 'Run'] (1, 50, 10, 10, 1)
...
```

## ~~Suggestions~~

~~Adding the following 3 lines would fix the problem **for BatchRunnerMP**~~ (see next comment)

```python
def _prepare_report_table(self, vars_dict, extra_cols=None):
    """
    Creates a dataframe from collected records and sorts it using 'Run'
    column as a key.
    """
    extra_cols = ["Run"] + (extra_cols or [])
    index_cols = []
    if self.parameters_list:
        index_cols = list(self.parameters_list[0].keys())

    # MY 2 CENTS ######################################
    if self.fixed_parameters:                                  # NEW
        fixed_param_cols = list(self.fixed_parameters.keys())  # NEW
        index_cols += fixed_param_cols                         # NEW
    #  ################################################

    index_cols += extra_cols
    .
    .
    .
    return ordered
```

which would print the following

```
['wealth_ini', 'N', 'width', 'height', 'Run'] (1, 20, 10, 10, 1)
['wealth_ini', 'N', 'width', 'height', 'Run'] (1, 20, 10, 10, 0)
['wealth_ini', 'N', 'width', 'height', 'Run'] (1, 20, 10, 10, 3)
['wealth_ini', 'N', 'width', 'height', 'Run'] (1, 20, 10, 10, 4)
['wealth_ini', 'N', 'width', 'height', 'Run'] (1, 50, 10, 10, 0)
['wealth_ini', 'N', 'width', 'height', 'Run'] (5, 20, 10, 10, 1)
['wealth_ini', 'N', 'width', 'height', 'Run'] (5, 20, 10, 10, 2)
['wealth_ini', 'N', 'width', 'height', 'Run'] (1, 50, 10, 10, 4)
['wealth_ini', 'N', 'width', 'height', 'Run'] (1, 50, 10, 10, 3)
['wealth_ini', 'N', 'width', 'height', 'Run'] (5, 20, 10, 10, 0)
...
```

## Limits

However, this solution would **cause problems with `BatchRunner`**. Because in this case, model_vars only include **`variable_parameters`** and the iterable **`Run`** (and not the fixed_parameters).

Indeed, it would print:

```python
['wealth_ini', 'N', 'width', 'height', 'Run'] (1, 20, 0)
['wealth_ini', 'N', 'width', 'height', 'Run'] (1, 20, 1)
['wealth_ini', 'N', 'width', 'height', 'Run'] (1, 20, 2)
['wealth_ini', 'N', 'width', 'height', 'Run'] (1, 20, 3)
['wealth_ini', 'N', 'width', 'height', 'Run'] (1, 20, 4)
['wealth_ini', 'N', 'width', 'height', 'Run'] (1, 50, 5)
['wealth_ini', 'N', 'width', 'height', 'Run'] (1, 50, 6)
['wealth_ini', 'N', 'width', 'height', 'Run'] (1, 50, 7)
['wealth_ini', 'N', 'width', 'height', 'Run'] (1, 50, 8)
['wealth_ini', 'N', 'width', 'height', 'Run'] (1, 50, 9)
['wealth_ini', 'N', 'width', 'height', 'Run'] (5, 20, 10)
...
```

**Comment by adrien-perello** [comment]
Actually my suggestion does not work. Here is what happened when I look at the average / deviation over 20 runs in series (BatchRunner)


![Screenshot from 2021-11-15 10-20-04](https://user-images.githubusercontent.com/76055749/141755692-94ab6289-6bce-44ad-9a2f-63f1600bd80b.png)

And here is what happened with the result from a parallel batch (BatchRunnerMP)


![Screenshot from 2021-11-15 10-20-15](https://user-images.githubusercontent.com/76055749/141755688-8b3abeb0-6fa9-4a42-8969-6f6d4536cf0b.png)

So somehow, the results get mixed up. If anyone has any idea how to solve this (I'm trying to generate an xarray from the [timeseries contained in DataCollectors](https://github.com/projectmesa/mesa/issues/186#issuecomment-966141870)

Example of desired result:
![Screenshot from 2021-11-15 10-32-26](https://user-images.githubusercontent.com/76055749/141757448-bc877a60-1617-4f24-baee-8b30b77e7962.png)

**Comment by adrien-perello** [comment]
Another issue also seems to rise when **`batch_run`** is only used with fixed parameters and no variable parameters.

This is the helper functions I currently use:

### To generate a clean dataframe:

```python
def generate_df(batch_run):
    # get dataframe from batch_run
    if isinstance(batch_run, BatchRunnerMP) or not batch_run.parameters_list:
        run_data = to_dataframe(batch_run)
    else:
        run_data = batch_run.get_model_vars_dataframe()
        # convert 'Run' number (=unique id) to iteration number
        # (for use as coordinate of future xarray)
        run_data["Run"] = run_data["Run"] % batch_run.iterations


    # drop fixed parameters (unecessary)
    fixed_params_idx = list(batch_run.fixed_parameters.keys())
    run_data.drop(fixed_params_idx, axis=1, inplace=True)
    # keep track of columns that will become indexes
    # (= all but datacollectors)
    indexes = list(run_data.drop("datacollectors", axis=1).columns)
    # unnest data contained in datacollectors into new columns
    run_data = unnest_records(run_data, "datacollectors")
    # keep track of new columns indexes (= variable names)
    variables = list(set(run_data.columns) - set(indexes))
    # reset index
    run_data.set_index(indexes, inplace=True, drop=True)
    # unnest values inside new columns (= dict) 
    return unnest_columns(run_data, variables)
```

### To generate a xarray:

```python
def generate_xr(batch_run):
    # generate cleaned dataframe
    df = generate_df(batch_run)
    #convert df to xarray
    da = df.to_xarray()
    # set fixed params as attributes rather than coords
    da.attrs = batch_run.fixed_parameters
    return da
```


### Other helper functions

```python
def to_dataframe(batch_run): # (self, vars_dict, extra_cols=None)
    extra_cols = ["Run"] # + (extra_cols or [])
    index_cols = []
    if batch_run.parameters_list:
        index_cols = list(batch_run.parameters_list[0].keys())
    if batch_run.fixed_parameters:
        fixed_param_cols = list(batch_run.fixed_parameters.keys())
        index_cols += fixed_param_cols
    index_cols += extra_cols
    records = []
    for param_key, values in batch_run.model_vars.items(): # self.vars_dict
        record = dict(zip(index_cols, param_key))
        record.update(values)
        records.append(record)
    df = pd.DataFrame(records)
    return df


def unnest_records(df, column):
    df_expanded = pd.DataFrame.from_records(df[column])
    df_final = pd.concat([df, df_expanded], axis=1)
    return df_final.drop(column, axis=1)


def unnest_columns(df, columns):
    return (pd.concat(
            [pd.DataFrame(df[x].tolist(), index=df.index)
                for x in columns],
            axis=1,
            keys=columns,
        )
        .stack()
        .rename_axis(index={None: "Step"})
    )

```

**Comment by tpike3** [comment]
@adrien-perello Thanks for all your work on this.  Could you please confirm the version of Mesa you are running?
@jackiekazil I think we need to @Corvince's [pull request](https://github.com/projectmesa/mesa/pull/924) so we can figure out if datacollector issues still remain

**Comment by adrien-perello** [comment]
I'm using Mesa 0.8.9

**Comment by tpike3** [comment]
@adrien-perello  this should be resolved with release 0.9.0; please let us know if you have any issues
